### PR TITLE
CDAP-12135 Support Wildcard grant/revoke

### PR DESCRIPTION
**Pull request by rohitsinha54**
- Add support for wildcard for grants and revoke in CDAP authorization
- Introduces Authorizable which is an representation of EntityId but also supports wildcards.
- Removes enforcement for grant/revoke

Build: https://builds.cask.co/browse/CDAP-RUT1261-3

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java [hunk 1]
```diff
@@ -231,22 +231,14 @@ private void ensureSecurityEnabled() throws FeatureDisabledException {
     }
   }
 
-  private void verifyAuthRequest(AuthorizationRequest request) throws BadRequestException, NotFoundException {
-    if (request == null) {
-      throw new BadRequestException("Missing request body");
-    }
-    EntityId entity = request.getEntity();
-    entityExistenceVerifier.ensureExists(entity);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java:239]: `entityExistenceVerifier` can be removed from the class

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java [hunk 2]
```diff
@@ -105,12 +104,12 @@ public void grant(HttpRequest httpRequest, HttpResponder httpResponder) throws E
...
 
     GrantRequest request = parseBody(httpRequest, GrantRequest.class);
-    verifyAuthRequest(request);
+    if (request == null) {
+      throw new BadRequestException("Missing request body");
+    }
 
     Set<Action> actions = request.getActions() == null ? EnumSet.allOf(Action.class) : request.getActions();
-    // enforce that the user granting access has admin privileges on the entity
-    authorizationEnforcer.enforce(request.getEntity(), authenticationContext.getPrincipal(), Action.ADMIN);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java:112]: `authorizationEnforcer` can be removed from the class

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java [hunk 3]
```diff
@@ -123,15 +122,16 @@ public void revoke(HttpRequest httpRequest, HttpResponder httpResponder) throws
     ensureSecurityEnabled();
 
     RevokeRequest request = parseBody(httpRequest, RevokeRequest.class);
-    verifyAuthRequest(request);
+    if (request == null) {
+      throw new BadRequestException("Missing request body");
+    }
 
     // enforce that the user revoking access has admin privileges on the entity
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java]: Comment needs to be removed

File: cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java [hunk 4]
```diff
@@ -381,13 +381,6 @@ public void testAuthorizationForPrivileges() throws Exception {
     String oldUser = getCurrentUser();
     setCurrentUser(alice.getName());
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java]: We can remove changing current user in the test since it is no longer  useful. Instead let's grant/list/reovke a wildcard privilege to make sure we don't throw any exceptions.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Class which represents entities on which privileges can be granted/revoked.
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: `Class to represent entities on which privileges can be granted/revoked.`

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+ * This class is based on {@link EntityId} but allows wildcards (* and ?) for wild card privileges.
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: `This class allows wildcard privileges (* and ?).`

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+ */
+public class Authorizable {
+  // represents the type of entity
+  private final EntityType entityType;
+  // represent the parts of the entity
+  private Map<EntityType, String> entityParts = new LinkedHashMap<>();
+
+  /**
+   * Constructs an {@link Authorizable} form the given entityString. The entityString must be an representation of
+   * entity similar to {@link EntityId#toString()} with the exception that the string can contains wildcards (? and *).
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: `Constructs an {@link Authorizable} from the given entityString. The entityString must be a representation of an entity similar to {@link EntityId#toString()} with the exception that the string can contain wildcards (? and *).

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+   * Note:
+   * <ol>
+   * <li>
+   * The only validation that this class performs on the entityString is that it checks if the string have enough
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: `if the string has enough`

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+   * valid parts for the given {@link #entityType}. It does not checks if the entity exists or not. This is
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: `It does not check if the entity exists or not.`

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+   * or applicationId which has version is passed to {@link Authorizable#fromEntityId(EntityId)} then the version
+   * will be silently dropped to construct the authorizable.
+   * </li>
+   * </ol>
+   *
+   * @param entityString the {@link EntityId#toString()} of the entity which may or may not contains wildcards(? or *)
+   */
+  public Authorizable(String entityString) {
+    if (entityString == null || entityString.isEmpty()) {
+      throw new IllegalArgumentException("A valid authorizable must be provided.");
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: Let's say `Null or empty entity string`

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    String[] typeAndId = entityString.split(EntityId.IDSTRING_TYPE_SEPARATOR, 2);
+    if (typeAndId.length != 2) {
+      throw new IllegalArgumentException(
+        String.format("Expected type separator '%s' to be in the ID string: %s", EntityId.IDSTRING_TYPE_SEPARATOR,
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: `Cannot extract entity type from %s`, entityString

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    String typeString = typeAndId[0];
+    EntityType type = EntityType.valueOf(typeString.toUpperCase());
+    String idString = typeAndId[1];
+    List<String> idParts = Arrays.asList(EntityId.IDSTRING_PART_SEPARATOR_PATTERN.split(idString));
+    checkParts(type, idParts, idParts.size() - 1);
+    entityType = type;
+  }
+
+  /**
+   * Created an {@link Authorizable} which represents the given entityId.
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: `Creates an ...`

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+      int versionEndIndex = entity.length();
+      if (entityId.getEntityType().equals(EntityType.PROGRAM)) {
+        // for programs versions doesn't end at the end of the entity string as there is program type and name
+        // afterwards
+        versionEndIndex = entity.lastIndexOf(EntityId.IDSTRING_PART_SEPARATOR,
+                                             entity.lastIndexOf(EntityId.IDSTRING_PART_SEPARATOR) - 1);
+      }
+      // remove the version from the entity string
+      String version = entity.substring(versionStartIndex, versionEndIndex);
+      entity = entity.replace(version, "");
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java:118]: It would be safer and simpler to split the string on `EntityId.IDSTRING_PART_SEPARATOR`, drop the index for the version, and then call `checkParts` on the remaining parts.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+  /**
+   * @return the type of the entity which is represented by this authorizable
+   */
+  public EntityType getEntityType() {
+    return entityType;
+  }
+
+  /**
+   * @return a map which represents the parts of the entity in ordered fashion. The
+   * internal implementation used a {@link LinkedHashMap} which guarantees that Stream comes after Namespace,
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: No need to expose the internal implementation, just say it will be ordered according to the hierarchy.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+   * Program comes after an Application etc. according to CDAP entity hierarchy.
+   */
+  public Map<EntityType, String> getEntityParts() {
+    return Collections.unmodifiableMap(entityParts);
+  }
+
+  /**
+   * @return a string representation of the authorizable which is compatible with {@link EntityId#toString()}.
+   * This is done to maintain backward compatibility for cli grant/revoke commands and consistency in representing
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: The second sentence can be removed.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+    Authorizable that = (Authorizable) o;
+    return entityType == that.entityType && entityParts.equals(that.entityParts);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(entityType, entityParts);
+  }
+
+  private void checkParts(EntityType entityType, List<String> parts, int curType) {
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: Changing `curType` to `index` will make the code more readable.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+   * and {@link co.cask.cdap.proto.id.ArtifactId}. If a version is included while construction an Authorizable
+   * through {@link #Authorizable(String)} an {@link IllegalArgumentException} will be thrown. If an artifactId
+   * or applicationId which has version is passed to {@link Authorizable#fromEntityId(EntityId)} then the version
+   * will be silently dropped to construct the authorizable.
+   * </li>
+   * </ol>
+   *
+   * @param entityString the {@link EntityId#toString()} of the entity which may or may not contains wildcards(? or *)
+   */
+  public Authorizable(String entityString) {
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: Let's change this to `fromString()` to be uniform with `fromEntityId()`

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * Tests {@link Authorizable}
+ */
+public class AuthorizableTest {
+
+  @Test
+  public void testFromEntity() {
```
Comment (poornachandra) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: Let's add tests for from string too

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    authorizable = Authorizable.fromEntityId(artifactId);
+    // drop the version while asserting
+    Assert.assertEquals(artifactId.toString().replace(".1.0-SNAPSHOT", ""), authorizable.toString());
+
+    ApplicationId appId = new ApplicationId("ns", "app", "1.0-SNAPSHOT");
+    authorizable = Authorizable.fromEntityId(appId);
+    // drop the version while asserting
+    Assert.assertEquals(appId.toString().replace(".1.0-SNAPSHOT", ""), authorizable.toString());
+
+    programId = appId.program(ProgramType.MAPREDUCE, "prog");
```
Comment (poornachandra) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: Let's use wildcards to make sure we can parse them correctly.

File: cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java [hunk 20]
```diff
@@ -33,28 +34,63 @@
    * @param entity the {@link EntityId} to whom {@link Action actions} are to be granted
    * @param principal the {@link Principal} that performs the actions. This could be a user, or role
    * @param actions the set of {@link Action actions} to grant.
+   * @deprecated As of release 4.3.0, replaced by {@link #grant(Authorizable, Principal, Set)}
    */
+  @Deprecated
   void grant(EntityId entity, Principal principal, Set<Action> actions) throws Exception;
 
   /**
+   * Grants a {@link Principal} authorization to perform a set of {@link Action actions} on an {@link EntityId}.
```
Comment (poornachandra) [inline cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java]: `Grants a {@link Principal} authorization to perform a set of {@link Action actions} on {@link EntityId}s represented by the {@link Authorizable}`

File: cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java [hunk 21]
```diff
@@ -33,28 +34,63 @@
...
+   * Note: this grant is used to support wildcard privilege management
+   *
+   * @param authorizable the {@link Authorizable} to whom {@link Action actions} are to be granted
```
Comment (poornachandra) [inline cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java]: `The {@link Authorizable} on which the {@link Action actions} are to be granted.

File: cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java [hunk 22]
```diff
@@ -33,28 +34,63 @@
...
    * @param principal the {@link Principal} that performs the actions. This could be a user, group or role
    * @param actions the set of {@link Action actions} to revoke
+   * @deprecated As of release 4.3.0, replaced by {@link #revoke(Authorizable, Principal, Set)}
    */
+  @Deprecated
   void revoke(EntityId entity, Principal principal, Set<Action> actions) throws Exception;
 
   /**
+   * Revokes a {@link Principal principal's} authorization to perform a set of {@link Action actions} on
+   * an {@link EntityId}.
```
Comment (poornachandra) [inline cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java]: Same changes here

File: cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java [hunk 23]
```diff
@@ -33,28 +34,63 @@
...
    *
    * @param entity the {@link EntityId} on which all {@link Action actions} are to be revoked
+   * @deprecated As of release 4.3.0, replaced by {@link #revoke(Authorizable)}
    */
+  @Deprecated
   void revoke(EntityId entity) throws Exception;
 
   /**
+   * Revokes all {@link Principal principals'} authorization to perform any {@link Action} on the given
+   * {@link EntityId}.
```
Comment (poornachandra) [inline cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/PrivilegesManager.java]: Same here

File: cdap-security/src/main/java/co/cask/cdap/proto/security/AuthorizationPrivilege.java [hunk 24]
```diff
@@ -17,48 +17,60 @@
...
+    return action;
+  }
+
   @Override
   public boolean equals(Object o) {
     if (this == o) {
       return true;
     }
-    if (o == null || getClass() != o.getClass()) {
+    if (o == null || getClass() != o.getClass())  {
```
Comment (poornachandra) [inline cdap-security/src/main/java/co/cask/cdap/proto/security/AuthorizationPrivilege.java]: Extra space

**Review by poornachandra — COMMENTED** [review]
@rohitsinha54 The changes look good, a few minor comments.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java (hunk 13 above)
Comment (rohitsinha54) [reply cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java:118]: That will not work as versions can have ```.``` which is also the EntityId.IDSTRING_PART_SEPARATOR and will mess up the split and part checks.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java (hunk 14 above)
Comment (rohitsinha54) [reply cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: 👍

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java (hunk 18 above)
Comment (rohitsinha54) [reply cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: this internally tests that since the fromEntity calls fromString

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java (hunk 19 above)
Comment (rohitsinha54) [reply cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: 👍

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java (hunk 18 above)
Comment (rohitsinha54) [reply cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: also added one with *

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java (hunk 3 above)
Comment (poornachandra) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AuthorizationHandler.java]: Comment is not addressed

File: cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java [hunk 25]
```diff
@@ -416,6 +397,33 @@ public void testRevokeAllOnNonExistingEntity()
     client.revoke(Ids.namespace("ns3"));
   }
 
+  @Test
+  public void testWildCardEntities() throws Exception {
+    // tin-mem authorizer does not support wildcard privileges so we can't actually enforce after grant.
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java]: in-mem

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 26]
```diff
@@ -27,57 +27,68 @@
...
+ * Class to represent entities on which privileges can be granted/revoked.
+ * This class allows wildcard privileges (* and ?).
  */
 public class Authorizable {
   // represents the type of entity
-  private final EntityType entityType;
+  private EntityType entityType;
   // represent the parts of the entity
-  private Map<EntityType, String> entityParts = new LinkedHashMap<>();
+  private Map<EntityType, String> entityParts;
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: This can also be final, right?

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java [hunk 27]
```diff
@@ -27,57 +27,68 @@
...
-    List<String> idParts = Arrays.asList(EntityId.IDSTRING_PART_SEPARATOR_PATTERN.split(idString));
-    checkParts(type, idParts, idParts.size() - 1);
-    entityType = type;
+    List<String> idParts;
+    if (type != EntityType.KERBEROSPRINCIPAL) {
+      // kerberos principal might contain . which is also EntityId.IDSTRING_PART_SEPARATOR_PATTERN so don't split for
+      // them and also principal doesn't have other parts so just initialize
+      idParts = Arrays.asList(EntityId.IDSTRING_PART_SEPARATOR_PATTERN.split(idString));
+    } else {
+      idParts = Collections.unmodifiableList(Collections.singletonList(idString));
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: Singleton list is already unmodifiable

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java (hunk 13 above)
Comment (poornachandra) [reply cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java:118]: Right - then split only the artifact, application and program strings like you are doing now. I think splitting the string and dropping the version using the index will make the code more readable, and it will be easy to spot the errors.

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java [hunk 28]
```diff
@@ -50,10 +63,21 @@ public void testFromEntity() {
...
+
     programId = appId.program(ProgramType.MAPREDUCE, "prog");
-    System.out.println(programId.toString());
     authorizable = Authorizable.fromEntityId(programId);
     // drop the version while asserting
     Assert.assertEquals(programId.toString().replace(".1.0-SNAPSHOT", ""), authorizable.toString());
+
+    KerberosPrincipalId kerberosPrincipalId = new KerberosPrincipalId("eve/host1.com@domain.net");
+    authorizable = Authorizable.fromEntityId(kerberosPrincipalId);
+    // drop the version while asserting
```
Comment (poornachandra) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: Looks like an unrelated comment.

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java [hunk 29]
```diff
@@ -31,15 +32,27 @@
...
+      Assert.fail();
+    } catch (UnsupportedOperationException e) {
+      // expected
+    }
+
     NamespaceId namespaceId = new NamespaceId("ns");
     authorizable = Authorizable.fromEntityId(namespaceId);
     Assert.assertEquals(namespaceId.toString(), authorizable.toString());
 
+    String wildcardNs = namespaceId.toString() + "*";
```
Comment (poornachandra) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: Let's use `?` here

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java [hunk 30]
```diff
@@ -31,15 +32,27 @@
   @Test
   public void testFromEntity() {
```
Comment (poornachandra) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java]: Unless we have the conversion tests elsewhere, let's add conversion tests for the other entity types here. Also would be good to add the wildcards in the middle of the entity name, in the parent entity name, etc. to make sure all possible cases work.

File: cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java [hunk 31]
```diff
@@ -416,6 +397,33 @@ public void testRevokeAllOnNonExistingEntity()
...
+    // this test is just to check that our grant apis work fine. After grant we list privileges to check for existence
+    StreamId streamId = new StreamId("ns", "wildcard");
+    String wildcardEntityStar = streamId.toString().replace("wildcard", "*");
+    client.grant(Authorizable.fromString(wildcardEntityStar), admin, ImmutableSet.of(Action.ADMIN));
+    String wildcardEntityQuestion = streamId.toString().replace("wildcard", "someSt?");
+    client.grant(Authorizable.fromString(wildcardEntityQuestion), admin, ImmutableSet.of(Action.ADMIN));
+
+    Set<Privilege> privileges = client.listPrivileges(admin);
```
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java:411]: Just to confirm - as a user, I can directly give the authorizable from the output of the `listPrivileges` method to the `revoke` method, right? Just want to make sure that output of the `listPrivileges` is in the same format as the input of `revoke` or `grant` method.

**Review by poornachandra — COMMENTED** [review]
@rohitsinha54 A few more comments

File: cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java (hunk 31 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java:411]: Yes modifies the  test to test this scenario. See line 422

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java (hunk 27 above)
Comment (rohitsinha54) [reply cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java]: 👍

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java (hunk 13 above)
Comment (rohitsinha54) [reply cdap-proto/src/main/java/co/cask/cdap/proto/security/Authorizable.java:118]: I am not able to understand this comment. If by index it means position then that's what being done now. 
If by index it means index or version after split that looks brittle since versions can have 1+ . in it. So currently we take elements from beginning and end which we want and leave the residue as version.
We decided to discuss this in person tomorrow.

File: cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java (hunk 31 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java:411]: 👍

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java [hunk 32]
```diff
@@ -69,15 +91,47 @@ public void testFromEntity() {
...
-    authorizable = Authorizable.fromEntityId(kerberosPrincipalId);
-    // drop the version while asserting
+    Authorizable authorizable = Authorizable.fromEntityId(kerberosPrincipalId);
     Assert.assertEquals(kerberosPrincipalId.toString(), authorizable.toString());
   }
+
+  @Test
+  public void testStream() {
+    StreamId streamId = new StreamId("ns", "stream");
+    Authorizable authorizable = Authorizable.fromEntityId(streamId);
```
Comment (poornachandra) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/AuthorizableTest.java:110]: >Unless we have the conversion tests elsewhere, let's add conversion tests for the other entity types here. Also would be good to add the wildcards in the middle of the entity name, in the parent entity name, etc. to make sure all possible cases work.

Missed adding wildcard in the middle of the entity name, in parent entity name, etc.

**Review by poornachandra — COMMENTED** [review]
Just one more comment needs to be addressed

**Review by poornachandra — APPROVED** [review]

**Review by poornachandra — APPROVED** [review]
LGTM, but wait until https://github.com/caskdata/cdap-security-extn/pull/144 is merged
